File: tests/omp/collapse.ml
open Optitrust
open Target

(* TODO: Fix the issue with indices for simple loops *)
let _ = Run.script_cpp (fun _ ->
  !! Omp.for_ [Collapse 2; Private ["i"; "k"; "j"]] [tBefore; cFunDef "sub1"; cFor_c "k"];
  !! Omp.parallel [] [tAfter; cFunDef "test"; cVarDef "j"];
  (*!! Omp.for_ [Collapse 2; LastPrivate ["jlast"; "klast"]] [tBefore; cFunDef "test"; cFor_c "k"];
  !! Omp.single [] [tAfter; cFunDef "test"; cFor_c "k"];
  !! Omp.parallel [NumThreads 2] [tAfter; cFunDef "sub2"; cVarDef "j"];
  !! Omp.for_ [Collapse 2; Ordered_c 0; Private ["j"; "k"] ; Schedule (Static, "3")] [tBefore; cFunDef "sub2"; cFor_c "k"]; *)
  (* !! Omp.ordered [] [tBefore; cFunDef "sub2"; cFun "printf"]; *) (* TODO: Fix the issue when matching printf call *)
)

File: tests/omp/collapse.cpp
#include <stdio.h>
#include <omp.h>

void bar(float *a, int i, int j, int k);
int kl, ku, ks, jl, ju, js, il, iu,is;
void sub1(float *a)
{
  for (int k=kl; k<=ku; k+=ks)
    for (int j=jl; j<=ju; j+=js)
      for (int i=il; i<=iu; i+=is)
        bar(a,i,j,k);
}

void test()
{
  int jlast, klast;
  {
    for (int k=1; k<=2; k++)
    for (int j=1; j<=3; j++)
  {
    jlast=j;
    klast=k;
  }
    printf("%d %d\n", klast, jlast);
  }
}

void work(int a, int j, int k);
void sub2()
{
   int j, k, a;
   {
      for (k=1; k<=3; k++)
         for (j=1; j<=2; j++)
         {
            printf("%d %d %d\n", omp_get_thread_num(), k, j);
            /* end ordered */
            work(a,j,k);
         }
   }
}

